### PR TITLE
Modify Add feature

### LINKED ISSUE #64: Modify Add feature
add items with their expiry dates (optional), quantity and current location

### PR DESCRIPTION
**Pull request by kumsssss**
1. Edit the Add Command method.
2. Make expiryDate optional for Add command.
3. Fix tests that went wrong after updating feature.
4. Update UserGuide for Add feature.
5. Add test case for Add.

Closes #64.

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -59,8 +59,14 @@ Adds an item to the inventory.
...
 Format: `add n/ITEM_NAME l/LOCATION q/QUANTITY [e/EXPIRY_DATE] [t/TAG]…​`
 
-<div markdown="span" class="alert alert-primary">:bulb: **Tip:**
+<div markdown="span" class="alert alert-primary">:bulb: **Tip:** 
 An item can have any number of tags (including 0)
+
+</div>
+
+<div markdown="span" class="alert alert-primary">:bulb: **Tip:** 
+An item can also not have any expiry date.
```
Comment (JayChenYJ) [inline docs/UserGuide.md]: Perhaps its better to say 'Expiry date of an item is optional'

File: src/main/java/seedu/storemando/model/item/ExpiryDate.java [hunk 2]
```diff
@@ -37,21 +48,42 @@
...
 
+    public LocalDate getDate() {
+        return this.date;
+    }
+
     /**
      * Returns if a given string is a valid expiryDate.
      */
     public static boolean isValidExpiryDate(String test) {
-        return test.matches(VALIDATION_REGEX);
```
Comment (chewwh09) [inline src/main/java/seedu/storemando/model/item/ExpiryDate.java]: Why did you remove this check?

File: src/main/java/seedu/storemando/model/item/Quantity.java [hunk 3]
```diff
@@ -31,7 +31,13 @@ public Quantity(String quantity) {
      * Returns true if a given string is a valid quantity number.
      */
     public static boolean isValidQuantity(String test) {
-        return test.matches(VALIDATION_REGEX);
+        if (test.matches(VALIDATION_REGEX)) {
```
Comment (chewwh09) [inline src/main/java/seedu/storemando/model/item/Quantity.java]: I think if you can avoid nesting if-else it would be great. So you can do
(test.matches(VALIDATION_REGEX) && value > 0) {
return true;
}

It increases readability as well.

**Review by chewwh09 — COMMENTED** [review]
Just nitpick a section of your code and require clarification on why you remove a check condition.

File: src/main/java/seedu/storemando/model/item/ExpiryDate.java [hunk 4]
```diff
@@ -3,31 +3,42 @@
...
+    /*private static final String LOCAL_PART_REGEX = "^[\\w" + SPECIAL_CHARACTERS + "]+";
     private static final String DOMAIN_FIRST_CHARACTER_REGEX = "[^\\W_]"; // alphanumeric characters except underscore
     private static final String DOMAIN_MIDDLE_REGEX = "[a-zA-Z0-9.-]*"; // alphanumeric, period and hyphen
     private static final String DOMAIN_LAST_CHARACTER_REGEX = "[^\\W_]$";
     public static final String VALIDATION_REGEX = LOCAL_PART_REGEX + "@"
         + DOMAIN_FIRST_CHARACTER_REGEX + DOMAIN_MIDDLE_REGEX + DOMAIN_LAST_CHARACTER_REGEX;
+     */
 
     public final String value;
+    public final LocalDate date;
```
Comment (Md-Fazil) [inline src/main/java/seedu/storemando/model/item/ExpiryDate.java]: Would it be better to name the field expiryDate to date? I think it would not be as ambiguous and would be easier to understand.

File: src/main/java/seedu/storemando/model/item/ExpiryDate.java (hunk 2 above)
Comment (Md-Fazil) [reply src/main/java/seedu/storemando/model/item/ExpiryDate.java]: Instead of first checking if the expiry date format is correct using regex, he parses the input right away and catches DateTimeParseException to determine if the specified input is a valid expiry date. I think it will still work as intended.

File: src/main/java/seedu/storemando/model/item/Item.java [hunk 5]
```diff
@@ -28,7 +28,7 @@
      * Every field must be present and not null.
      */
     public Item(ItemName name, Quantity quantity, ExpiryDate expiryDate, Location location, Set<Tag> tags) {
-        requireAllNonNull(name, quantity, expiryDate, location, tags);
+        requireAllNonNull(name, quantity, location, tags);
```
Comment (Md-Fazil) [inline src/main/java/seedu/storemando/model/item/Item.java]: Should this nonNull check for expiryDate be removed?

File: src/main/java/seedu/storemando/model/item/ExpiryDate.java (hunk 2 above)
Comment (kumsssss) [reply src/main/java/seedu/storemando/model/item/ExpiryDate.java]: I felt like the VALIDATION_REGEX does not help to check if the date given by the user is a valid date. So, I used LocalDate instead to check if the date is valid, which can be seen in the subsequent lines. Hope this clears things.

**Review by Md-Fazil — COMMENTED** [review]
Some nits and grits to fix before merging.

File: src/main/java/seedu/storemando/model/item/Quantity.java (hunk 3 above)
Comment (kumsssss) [reply src/main/java/seedu/storemando/model/item/Quantity.java:34]: Alright, I will update it respectively.

File: src/main/java/seedu/storemando/model/item/Item.java (hunk 5 above)
Comment (kumsssss) [reply src/main/java/seedu/storemando/model/item/Item.java]: I will add it back. expiryDate will not be null anymore.

File: docs/UserGuide.md (hunk 1 above)
Comment (github-amanda) [reply docs/UserGuide.md]: Agreed

File: src/main/java/seedu/storemando/model/item/Quantity.java (hunk 3 above)
Comment (kumsssss) [reply src/main/java/seedu/storemando/model/item/Quantity.java]: I chose to do it this way as I cannot guarantee Long.parseLong(test) will not throw a NumberFormatException. So, I made sure test is made of digits before I check if it is a non-zero value.

File: docs/UserGuide.md (hunk 1 above)
Comment (kumsssss) [reply docs/UserGuide.md]: Ok. Changed as stated.

File: src/main/java/seedu/storemando/model/item/ExpiryDate.java (hunk 4 above)
Comment (kumsssss) [reply src/main/java/seedu/storemando/model/item/ExpiryDate.java]: alright!

File: src/main/java/seedu/storemando/model/item/Quantity.java (hunk 3 above)
Comment (chewwh09) [reply src/main/java/seedu/storemando/model/item/Quantity.java]: But as long as your Long.parseLong(test) is the second condition to test.matches(VALIDATION_REGEX), it will be fine due to short-circuit.

Comment (kumsssss) [reply src/main/java/seedu/storemando/model/item/Quantity.java]: ohh ok i get it. I'll change it. Thanks!

File: src/main/java/seedu/storemando/model/item/ExpiryDate.java [hunk 6]
```diff
@@ -49,39 +44,39 @@ public ExpiryDate(String expiryDate) {
...
-            return true;
-        }
         try {
+            if (test.equals(NO_EXPIRY_DATE)) {
+                return true;
+            }
             LocalDate.parse(test);
             return true;
-        } catch (DateTimeParseException e) {
+        } catch (DateTimeParseException | NullPointerException e) {
```
Comment (JayChenYJ) [inline src/main/java/seedu/storemando/model/item/ExpiryDate.java]: Why is there an additional catch NullPointerException?

File: src/main/java/seedu/storemando/model/item/Item.java [hunk 7]
```diff
@@ -28,7 +28,7 @@
      * Every field must be present and not null.
      */
     public Item(ItemName name, Quantity quantity, ExpiryDate expiryDate, Location location, Set<Tag> tags) {
-        requireAllNonNull(name, quantity, location, tags);
+        requireAllNonNull(name, quantity, expiryDate, location, tags);
```
Comment (JayChenYJ) [inline src/main/java/seedu/storemando/model/item/Item.java]: understand that for this file the sequence in which what is placed first does not matter, but for readability and consistency of our code base we should perhaps standardise it, following the google docs sequence

File: src/main/java/seedu/storemando/model/item/Quantity.java [hunk 8]
```diff
@@ -12,8 +12,9 @@
 
 
     public static final String MESSAGE_CONSTRAINTS =
-        "Quantity numbers should only contain numbers, and it should be at least 1 digit long";
-    public static final String VALIDATION_REGEX = "\\d{1,}";
+        "Quantity numbers should only contain numbers, and it should be a positive integer greater than 0";
```
Comment (JayChenYJ) [inline src/main/java/seedu/storemando/model/item/Quantity.java]: I think "Quantity should only contain numbers..." will suffice, can remove the "numbers" after Quantity

File: src/main/java/seedu/storemando/model/item/ExpiryDate.java [hunk 9]
```diff
@@ -27,10 +29,11 @@
         + "    - be within the range [01, 12]\n"
         + "4. DD must:\n"
         + "    - be exactly 2 digits long\n"
-        + "    - be a valid date\n";
+        + "    - be a valid expiryDate\n";
```
Comment (github-amanda) [inline src/main/java/seedu/storemando/model/item/ExpiryDate.java]: Would "expiry date" be better than "expiryDate"?

**Review by github-amanda — APPROVED** [review]
Minor change: may be better to use "expiry date" instead of "expiryDate". Otherwise, LGTM!
